avro-fastserde/src/main/java/com/linkedin/avro/fastserde/FastDeserializerGeneratorBase.java

```diff
@@ -0,0 +1,303 @@
+import java.util.Collections;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ThreadLocalRandom;
+import javax.tools.JavaCompiler;
+import javax.tools.ToolProvider;
+import org.apache.avro.Schema;
+import org.apache.avro.io.parsing.Symbol;
+import org.apache.log4j.Logger;
+
+
+public abstract class FastDeserializerGeneratorBase<T> {
+  public static final String GENERATED_PACKAGE_NAME = "com.linkedin.avro.fastserde.deserialization.generated";
+  public static final String GENERATED_SOURCES_PATH = "/com/linkedin/avro/fastserde/deserialization/generated/";
+  protected static final Symbol EMPTY_SYMBOL = new Symbol(Symbol.Kind.TERMINAL, new Symbol[]{}) {
+  };
+  protected static final Symbol END_SYMBOL = new Symbol(Symbol.Kind.TERMINAL, new Symbol[]{}) {
+  };
+  private static final Logger LOGGER = Logger.getLogger(FastDeserializerGeneratorBase.class);
+  private static final Map<Schema, Integer> SCHEMA_IDS_CACHE = new ConcurrentHashMap<>();
+  private static final HashFunction HASH_FUNCTION = Hashing.murmur3_128();
+  protected final Schema writer;
+  protected final Schema reader;
+  protected JCodeModel codeModel;
+  protected JDefinedClass deserializerClass;
+  private File destination;
+  private ClassLoader classLoader;
+  private String compileClassPath;
+
+  FastDeserializerGeneratorBase(Schema writer, Schema reader, File destination, ClassLoader classLoader,
+      String compileClassPath) {
+    this.writer = writer;
+    this.reader = reader;
+    this.destination = destination;
+    this.classLoader = classLoader;
+    this.compileClassPath = compileClassPath;
+    codeModel = new JCodeModel();
+  }
+
+  protected static Symbol[] reverseSymbolArray(Symbol[] symbols) {
+    Symbol[] reversedSymbols = new Symbol[symbols.length];
+
+    for (int i = 0; i < symbols.length; i++) {
+      reversedSymbols[symbols.length - i - 1] = symbols[i];
+    }
+
+    return reversedSymbols;
+  }
+
+  public static String getClassName(Schema writerSchema, Schema readerSchema, String description) {
+    Integer writerSchemaId = Math.abs(getSchemaId(writerSchema));
+    Integer readerSchemaId = Math.abs(getSchemaId(readerSchema));
+    Schema.Type readerSchemaType = readerSchema.getType();
+    if (Schema.Type.RECORD.equals(readerSchemaType)) {
+      return readerSchema.getName() + description + "Deserializer" + writerSchemaId + "_" + readerSchemaId;
+    } else if (Schema.Type.ARRAY.equals(readerSchemaType)) {
+      return "Array" + description + "Deserializer" + writerSchemaId + "_" + readerSchemaId;
+    } else if (Schema.Type.MAP.equals(readerSchemaType)) {
+      return "Map" + description + "Deserializer" + writerSchemaId + "_" + readerSchemaId;
+    }
+    throw new FastDeserializerGeneratorException("Unsupported return type: " + readerSchema.getType());
+  }
+
+  protected static String getVariableName(String name) {
+    return name + nextRandomInt();
```

Comment:
i think a static final AtomicInteger + getAndIncrement() would lead to more readable (maybe even consistent) results and is safer

Reply:
Adopted the recommendations!